### PR TITLE
style: Document and force documentation in a big chunk of the style crate.

### PR DESCRIPTION
**Pull request by emilio**
Style no forced docs for the properties code and similar, but I ran out of time, and I think it's a nice improvement.

I'd appreciate a fast-ish turn-around time because this is pretty much prone to bitrot.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14801)

### CODE REVIEW

**Comment by emilio** [comment]
Mat said he was able to review this, thanks!

@bors-servo r? @mbrubeck

File: components/style/rule_tree/mod.rs [hunk 1]
```diff
@@ -412,14 +453,19 @@ impl StrongRuleNode {
...
     pub fn style_source(&self) -> Option<&StyleSource> {
         self.get().source.as_ref()
     }
 
+    /// Get the importance that this rule node represents.
     pub fn importance(&self) -> Importance {
         self.get().importance
     }
 
+    /// Get an iterator for this rule node and its ancestors including.
```
Comment (mbrubeck) [inline components/style/rule_tree/mod.rs]: "including" is dangling here. Should maybe be removed or changed to "inclusive."

File: components/style/rule_tree/mod.rs [hunk 2]
```diff
@@ -17,15 +20,40 @@ use std::sync::atomic::{AtomicPtr, AtomicUsize, Ordering};
 use stylesheets::StyleRule;
 use thread_state;
 
+/// The rule tree, the structure servo uses to preserve the results of selector
+/// matching.
+///
+/// This is organized as a tree of rules. When a node matches a set of rules,
+/// they're inserted in order in the tree, starting with the less specific one.
+///
+/// When a rule is inserted in the tree, other element may share the path up to
```
Comment (mbrubeck) [inline components/style/rule_tree/mod.rs]: s/element/elements/

File: components/style/rule_tree/mod.rs [hunk 3]
```diff
@@ -17,15 +20,40 @@ use std::sync::atomic::{AtomicPtr, AtomicUsize, Ordering};
...
+/// a given rule. If that's the case, we don't duplicate child nodes, but share
+/// them.
+///
+/// When the rule node refcount drops to zero, it doesn't get freed. It gets
+/// instead being put into a free list, and it is potentially GC'd after a
```
Comment (mbrubeck) [inline components/style/rule_tree/mod.rs]: "being" can be removed

**Review by mbrubeck — CHANGES_REQUESTED** [review]
Awesome work! See the comments for some very minor copy-editing. Can land with r=mbrubeck.

File: components/style/properties/helpers/animated_properties.mako.rs [hunk 4]
```diff
@@ -919,27 +945,36 @@ impl Interpolate for LengthOrNone {
...
+    /// A 2d matrix for interpolation.
     #[derive(Clone, Copy, Debug)]
     #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    #[allow(missing_docs)]
     pub struct InnerMatrix2D {
         pub m11: CSSFloat, pub m12: CSSFloat,
         pub m21: CSSFloat, pub m22: CSSFloat,
     }
 
+    /// A translation function.
```
Comment (Manishearth) [inline components/style/properties/helpers/animated_properties.mako.rs]: 2d

File: components/style/properties/properties.mako.rs [hunk 5]
```diff
@@ -158,7 +158,13 @@ pub mod shorthands {
     <%include file="/shorthand/text.mako.rs" />
 }
 
+/// A module with all the code related to animated properties.
+///
+/// This needs to be loaded at least after all longhand modules, given they
```
Comment (Manishearth) [inline components/style/properties/properties.mako.rs]: might want to clarify that "loaded" has to do with mako eval order

File: components/style/properties/helpers.mako.rs [hunk 6]
```diff
@@ -405,7 +413,11 @@
         }
 
         impl<'a> LonghandsToSerialize<'a> {
-            pub fn from_iter<I: Iterator<Item=&'a PropertyDeclaration>>(iter: I) -> Result<Self, ()> {
+            /// Maybe gets a serializable set of longhands given a set of
```
Comment (Manishearth) [inline components/style/properties/helpers.mako.rs]: "Tries to get"

File: components/style/properties/helpers.mako.rs [hunk 7]
```diff
@@ -477,7 +491,10 @@
         }
 
 
-        pub fn parse(context: &ParserContext, input: &mut Parser,
+        /// Parse the given shorthand and return the result in the
```
Comment (Manishearth) [inline components/style/properties/helpers.mako.rs]: "fill the result into"

File: components/style/properties/longhand/box.mako.rs [hunk 8]
```diff
@@ -275,44 +290,49 @@ ${helpers.single_keyword("overflow-x", "visible hidden scroll auto",
...
-  pub mod computed_value {
-      #[derive(Debug, Clone, Copy, PartialEq)]
-      #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-      pub struct T(pub super::super::overflow_x::computed_value::T);
-  }
 
-  impl ComputedValueAsSpecified for SpecifiedValue {}
+    /// The specified and computed value for overflow-y is a wrapper on top of
+    /// `overflow-x`, so we re-use the logic, but prevent errors from mistakenly
+    /// assign one to other.
```
Comment (Manishearth) [inline components/style/properties/longhand/box.mako.rs]: IMO we should actually be using the same value here. There's precedent for having typedefs.

File: components/style/properties/properties.mako.rs [hunk 9]
```diff
@@ -1821,16 +1845,22 @@ pub fn apply_declarations<'a, F, I>(viewport_size: Size2D<Au>,
...
 
     style
 }
 
+/// Modifies the style for an anonymous flow so it resets all its non-inherited
+/// style structs, and set their borders and outlines to zero.
+///
+/// Also, it gets a new display value, which is honored except when it's
+/// `inline`.
+/// values reseted new display, and also
```
Comment (Manishearth) [inline components/style/properties/properties.mako.rs]: and also...?

File: components/style/values/mod.rs [hunk 10]
```diff
@@ -46,17 +47,26 @@ macro_rules! define_numbered_css_keyword_enum {
 pub mod computed;
 pub mod specified;
 
+/// A CSS float value.
 pub type CSSFloat = f32;
 
+/// The default font size.
 pub const FONT_MEDIUM_PX: i32 = 16;
 
+/// A trait used to represent whether this value has viewport units.
```
Comment (Manishearth) [inline components/style/values/mod.rs]: s/represent/query

File: components/style/values/mod.rs [hunk 11]
```diff
@@ -46,17 +47,26 @@ macro_rules! define_numbered_css_keyword_enum {
...
 pub trait HasViewportPercentage {
+    /// Returns true if this value has viewport units.
     fn has_viewport_percentage(&self) -> bool;
 }
 
+/// A trait used as a marker to represent that a given value has no viewport
```
Comment (Manishearth) [inline components/style/values/mod.rs]: "given type may never contain viewport units"

**Review by Manishearth — COMMENTED** [review]
reviewed up till (not including) "style: Document a bit better the rule tree code." may pick up later if someone else doesn't.

**Comment by emilio** [comment]
Thanks for the review! I've addressed the copy-editing comments.

Going to land this as-is, document the pending stuff, then make `missing_docs` by default, probably leaving the properties module out for now.

@bors-servo r=mbrubeck,Manishearth,Wafflespeanut

**Pull request merged**